File: examples/biased-coin-flip/perturbation-pdf.py
from numpy import pi, sqrt, exp
from sys import argv, stdin, stdout, stderr

# Process arguments
if len(argv) != 2:
    stderr.write("Usage: {} STDEV\n"
            "Return probability densities of normal distributions centered \n"
            "on given parameters for the perturbed parameter \n"
            "and with standard deviation STDEV\n".format(argv[0]))

    exit(1)

stdev = float(argv[1])

# Read generation t and perturbed parameter from stdin
stderr.write("Enter t\n")
t = int(stdin.readline())

stderr.write("Enter q_perturbed\n")
q_perturbed = float(stdin.readline())

# Define normal pdf
normal_pdf = lambda mu, x: 1.0 / sqrt(2 * pi * stdev**2) * \
        exp( - (x - mu)**2 / (2.0 * stdev**2))

# Read parameters from stdin
stderr.write("Enter parameters\n")
q_array = []
for line in stdin:
    q_array.append(float(line))

# Return normal_pdf for every parameter
for q in q_array:
    pdf = normal_pdf(q, q_perturbed)
    stdout.write("{}\n".format(pdf))

File: examples/biased-coin-flip/perturbation-pdf.cc
#include <string>
#include <vector>
#include <iostream>

#include <math.h>

/** Return probabilility density function of normal distribution.
 *
 * @param mu  mean of normal distribution.
 * @param sigma  standard deviation of normal distribution.
 * @param x  position to evaluate distribution on.
 *
 * @return probability density of normal distribution at x.
 */
double normal_pdf(double mu, double sigma, double x)
{

    return exp( - (x - mu) * (x - mu) / ( 2.0 * sigma * sigma ) ) /
                 sqrt( 2.0 * M_PI * sigma * sigma );
}

int main(int argc, char *argv[])
{
    // Process arguments
    if (argc != 2)
    {
        std::cerr << "Usage: " << argv[0] << " STDEV\n"
            "Return probability densities of normal distributions centered \n"
            "on given parameters for the perturbed parameter \n"
            "and with standard deviation STDEV\n";
        return 1;
    }

    double stdev = std::stod(argv[1]);

    // Read generation t and parameter from stdin
    int t;
    std::cin >> t;

    double q_perturbed;
    std::cin >> q_perturbed;

    if (std::cin.fail())
    {
        std::cerr << "Error: could not read t or q from stdin\n";
        return 1;
    }

    // Return normal_pdf for every parameter
    std::cout.precision(17);
    std::vector<double> q_vec;
    double q;
    for (std::cin >> q; !std::cin.eof(); std::cin >> q)
    {
        if (std::cin.fail())
        {
            std::cerr << "Error: could not read parameter from stdin\n";
            return 1;
        }

        q_vec.push_back(q);
    }

    for (const double& q : q_vec)
    {
        double pdf = normal_pdf(q, stdev, q_perturbed);
        std::cout << pdf << std::endl;
    }

    return 0;
}
